### PR TITLE
GG-6918

### PR DESCRIPTION
**Pull request by szabolcs-szilagyi-hmrc**
Fix bug to not log html page when downstream returns it
improve error logging when fetching corporation tax id

https://jira.tools.tax.service.gov.uk/browse/GG-6918

### CODE REVIEW

File: app/uk/gov/hmrc/incorporatedentityidentification/httpparsers/GetCtReferenceHttpParser.scala [hunk 1]
```diff
@@ -18,28 +18,34 @@ package uk.gov.hmrc.incorporatedentityidentification.httpparsers
...
+            case Success(json) =>
+              (json \ ctutrKey).validate[String] match {
+                case JsSuccess(ctutr, _) =>
+                  Right(ctutr)
+                case JsError(errors) =>
+                  Left(new BadGatewayException(s"HoD returned a malformed JSON on $method <$url> errors: $errors"))
+              }
+            case Failure(exception) =>
+              Left(new BadGatewayException(s"HoD returned a malformed JSON on $method <$url> errors: ${exception.getMessage.take(40)}"))
           }
```
Comment (ModisR) [inline app/uk/gov/hmrc/incorporatedentityidentification/httpparsers/GetCtReferenceHttpParser.scala:43]: Lines 33-42 can be distilled down to:
```scala
val result = for {
  json <- Try(response.json).toEither.left.map(_.getMessage take 40)
  ctutr <- (json \ ctutrKey).validate[String].asEither.left.map(_.toString)
} yield ctutr

result.left.map { errors =>
  new BadGatewayException(s"HoD returned a malformed JSON on $method <$url> errors: $errors")
}
```

File: app/uk/gov/hmrc/incorporatedentityidentification/services/ValidateIncorporatedEntityDetailsService.scala [hunk 2]
```diff
@@ -17,22 +17,26 @@
...
-      case Some(retrievedCtUtr) => optCtUtr match {
-        case Some(`retrievedCtUtr`) => DetailsMatched
-        case Some(_) | None => DetailsMismatched
-      }
-      case None => DetailsNotFound
+      case Right(retrievedCtUtr) =>
+        optCtUtr match {
+          case Some(`retrievedCtUtr`) => DetailsMatched
+          case Some(_) | None         => DetailsMismatched
+        }
```
Comment (ModisR) [inline app/uk/gov/hmrc/incorporatedentityidentification/services/ValidateIncorporatedEntityDetailsService.scala:37]: Would it be clearer to write lines 34-37 as follows?
```scala
if( optCtUtr contains retrievedCtUtr ) DetailsMatched
else                                   DetailsMismatched
```

**Review by ModisR — DISMISSED** [review]

File: app/uk/gov/hmrc/incorporatedentityidentification/httpparsers/GetCtReferenceHttpParser.scala [hunk 3]
```diff
@@ -18,28 +18,34 @@ package uk.gov.hmrc.incorporatedentityidentification.httpparsers
...
+          Try(response.json) match {
+            case Success(json) =>
+              (json \ ctutrKey).validate[String] match {
+                case JsSuccess(ctutr, _) =>
+                  Right(ctutr)
+                case JsError(errors) =>
+                  Left(new BadGatewayException(s"HoD returned a malformed JSON on $method <$url> errors: $errors"))
+              }
+            case Failure(exception) =>
+              Left(new BadGatewayException(s"HoD returned a malformed JSON on $method <$url> errors: ${exception.getMessage.take(40)}"))
```
Comment (ModisR) [inline app/uk/gov/hmrc/incorporatedentityidentification/httpparsers/GetCtReferenceHttpParser.scala:42]: Lines 33-42 can be distilled down to:
```scala
val result = for {
  json <- Try(response.json).toEither.left.map(_.getMessage take 40)
  ctutr <- (json \ ctutrKey).validate[String].asEither.left.map(_.toString)
} yield ctutr

result.left.map { errors =>
  new BadGatewayException(s"HoD returned a malformed JSON on $method <$url> errors: $errors")
}
```

**Review by ModisR — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/incorporatedentityidentification/httpparsers/GetCtReferenceHttpParser.scala (hunk 1 above)
Comment (szabolcs-szilagyi-hmrc) [reply app/uk/gov/hmrc/incorporatedentityidentification/httpparsers/GetCtReferenceHttpParser.scala:39]: https://github.com/hmrc/incorporated-entity-identification/commit/8f360d7f92d41722310f374c3b8e08863af5579d

File: app/uk/gov/hmrc/incorporatedentityidentification/services/ValidateIncorporatedEntityDetailsService.scala (hunk 2 above)
Comment (szabolcs-szilagyi-hmrc) [reply app/uk/gov/hmrc/incorporatedentityidentification/services/ValidateIncorporatedEntityDetailsService.scala:37]: I agree, I did not want to change too much about the code but now there is two of us I'm happy to change it

File: app/uk/gov/hmrc/incorporatedentityidentification/services/JourneyIdGenerationService.scala [hunk 4]
```diff
@@ -21,6 +21,6 @@ import java.util.UUID
 import javax.inject.{Inject, Singleton}
 
 @Singleton
-class JourneyIdGenerationService @Inject()() {
+class JourneyIdGenerationService @Inject() () {
   def generateJourneyId(): String = UUID.randomUUID().toString
 }
```
[the remark this answers is not in the document]
Comment (szabolcs-szilagyi-hmrc) [reply app/uk/gov/hmrc/incorporatedentityidentification/services/JourneyIdGenerationService.scala:26]: Good suggestion but removing it is causing issues in other tests as this class is used to easily mock the random id generation. There is another way of doing it but that would require a bigger refactor which I think does not worth it just to remove a small class.

[the remark this answers is not in the document]
Comment (szabolcs-szilagyi-hmrc) [reply app/uk/gov/hmrc/incorporatedentityidentification/services/JourneyIdGenerationService.scala:26]: controllers.JourneyDataControllerISpec This is the test which is using it to mock the generated journey id

File: app/uk/gov/hmrc/incorporatedentityidentification/services/ValidateIncorporatedEntityDetailsService.scala (hunk 2 above)
Comment (szabolcs-szilagyi-hmrc) [reply app/uk/gov/hmrc/incorporatedentityidentification/services/ValidateIncorporatedEntityDetailsService.scala]: https://github.com/hmrc/incorporated-entity-identification/commit/fa036cd13df8b0feafbce960a22898ad5c4d823c

File: app/uk/gov/hmrc/incorporatedentityidentification/services/JourneyIdGenerationService.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (szabolcs-szilagyi-hmrc) [reply app/uk/gov/hmrc/incorporatedentityidentification/services/JourneyIdGenerationService.scala:26]: @ModisR What do you think?

[the remark this answers is not in the document]
Comment (ModisR) [reply app/uk/gov/hmrc/incorporatedentityidentification/services/JourneyIdGenerationService.scala:26]: I think that the ITs are written in a bad way - they're persisting session state between test cases, violating atomicity. However, you are right it's not worth untangling that at this point. Will approve PR.

**Review by ModisR — APPROVED** [review]

**Pull request merged**